File: src/2023/iter-4(~2023-10-31)/64. Minimum Path Sum/p2.py
from typing import List
from collections import deque

"""
Input: grid = [[1,3,1],[1,5,1],[4,2,1]]
Output: 7
Explanation: Because the path 1 → 3 → 1 → 1 → 1 minimizes the sum.
"""
"""
Constraints:

m == grid.length
n == grid[i].length
1 <= m, n <= 200
0 <= grid[i][j] <= 200
"""


# dp
def minPathSum(grid: List[List[int]]) -> int:
    m, n = len(grid), len(grid[0])

    for i in range(m):
        for j in range(n):
            if i == 0 and j == 0:
                continue

            left = float("inf")
            up = float("inf")
            upIdx = (i - 1, j)
            leftIdx = (i, j - 1)
            if upIdx[0] >= 0 and upIdx[0] <= m - 1:
                up = grid[i][j] + grid[upIdx[0]][upIdx[1]]
            if leftIdx[1] >= 0 and leftIdx[1] <= n - 1:
                left = grid[i][j] + grid[leftIdx[0]][leftIdx[1]]

            grid[i][j] = min(left, up)

    return grid[m - 1][n - 1]


print(minPathSum([[1, 3, 1], [1, 5, 1], [4, 2, 1]]))
print(minPathSum([[1, 2, 3], [4, 5, 6]]))

# https://leetcode.com/problems/minimum-path-sum/

File: src/2024/iter-8(~2024-02-29)/64. Minimum Path Sum/p2.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function (grid) {
  // only move down or right
  // 거꾸로 생각해서 left or up 으로 계산하기
  const m = grid.length;
  const n = grid[0].length;

  // prevRow
  // currRow
};

/**
 * Output: 7
  Explanation: Because the path 1 → 3 → 1 → 1 → 1 minimizes the sum.
 */
// console.log(
//   minPathSum([
//     [1, 3, 1],
//     [1, 5, 1],
//     [4, 2, 1],
//   ])
// );

console.log(
  minPathSum([
    [1, 2],
    [5, 6],
    [1, 1],
  ])
);
